### ISSUE TITLE
Natively Declarable Fast-path Functions 

### ISSUE THREAD

**Issue by eellison** [pytorch/pytorch#30774]
## 🚀 Feature

Fast-path functions such as `numel` are not natively registered to avoid dispatch overhead. Example PR here https://github.com/pytorch/pytorch/pull/27294 which removed `numel` from JIT-models potentially breaking BC.  In order for this function to be accessible in JIT we would have to duplicate it. 

It would be nice to have these functions declarable natively and shared across JIT / Pytorch eager.
  
Functions which fall under this category:
- `numel`
- `get_device`
- `is_contiguous`
- `storage_offset`
- `ndimension` 
- `dim`
- `is_cuda, is_sparse, is_mkldnn, is_quantized`

**Comment by gchanan** [comment]
maybe add a "final" specification somewhere.

**Comment by kurtamohler** [comment]
I'm interested in working on this because I want to learn more about JIT. But please excuse my lack of knowledge, I think I need to understand a few things more clearly before I could start working on this.

As far as I understand, when `numel` was removed from `native_functions.yaml`, it now no longer goes through the dispatch code, so it has better performance than it used to. Is that right?

The description mentions that this change to `numel` may have caused BC breaking issues. What exactly are these BC breaks? Is this because the dispatch code performs checks and possibly throws errors in some cases, so when the dispatch code was avoided, these checks are no longer being done?

The issue description mentions "In order for this function to be accessible in JIT we would have to duplicate it." But aren't these methods already accessible in JIT? We have unit tests that currently pass, like `test_script_get_device_cuda` and `test_ numel` in `test_jit.py`.

I'm not sure if I understand what "natively declarable" means. To me, it sounds like that would mean "declared in `native_functions.yaml`". But if we moved functions back to `native_functions.yaml`, then we would no longer be avoiding the dispatch code, so performance would get worse.

**Comment by ezyang** [comment]
I think there's a pretty substantial unknown design component for this, which may make it difficult for non-internals team to work on. I'm happy to answer your questions though, since they should be helpful for whoever eventually works on this.

> As far as I understand, when numel was removed from native_functions.yaml, it now no longer goes through the dispatch code, so it has better performance than it used to. Is that right?

Yes. numel is manually implemented in TensorBody.h like this:

```
  int64_t numel() const {
    return impl_->numel();
  }
```

if it had a declaration in `native_functions.yaml`, it would have gotten a code generated implementation that would have gone through codegen that would call the dispatcher.

The problem is that removing numel from `native_functions.yaml` ALSO disables a bunch of downstream codegen that would be pretty useful, including JIT codegen.

> The description mentions that this change to numel may have caused BC breaking issues. What exactly are these BC breaks? Is this because the dispatch code performs checks and possibly throws errors in some cases, so when the dispatch code was avoided, these checks are no longer being done?

No, it's simpler than that. Operators that don't exist in `native_functions.yaml` no longer automatically get parsed by JIT serialized models, so if they used numel they'd break (oops!) There are a bunch of operators which JIT manually implemented, and someone added numel after the fact in this PR: https://github.com/pytorch/pytorch/pull/36551/files

> The issue description mentions "In order for this function to be accessible in JIT we would have to duplicate it." But aren't these methods already accessible in JIT? We have unit tests that currently pass, like test_script_get_device_cuda and test_ numel in test_jit.py.

Yes. It works because we have duplicated it. 

AFAICT the amount of JIT specific duplication these days is not so bad, it's just the registration into the primitve ops (see PR above). But there's also a bunch of Python binding boilerplate you have to do as well, and it's all unnecessary because it should be the same as what the codegen would produce.

> I'm not sure if I understand what "natively declarable" means. To me, it sounds like that would mean "declared in native_functions.yaml". But if we moved functions back to native_functions.yaml, then we would no longer be avoiding the dispatch code, so performance would get worse.

Yes. So the crux of the issue is, define a function in `native_functions.yaml`, but then define some syntax to say "but don't use the dispatcher, and here is the real implementation to use." This is the design component that will need some work collectively with internals team.

cc @bdhirsh (this is not one of your starter projects, but it might be a good intermediate term project)

**Comment by kurtamohler** [comment]
Got it, thanks @ezyang. Since the design decisions need to be made internally, maybe I should put the quansight-nack label on this?

**Comment by ezyang** [comment]
Yes, unless perhaps if I've failed to dissuade you haha. (I'd be happy to advise this project if you did want to pick it up.)

**Comment by kurtamohler** [comment]
Well I would still be interested in it. But maybe I should wait a bit, learn more about the machinery involved in this issue over the next several weeks/months. If I think I've reached a sufficient understanding, and no one else has picked up this issue by then, maybe I'll pick it up. Also, if by any chance someone wanted to hash out the design decisions and didn't want to implement them for some reason, I'd be happy to do it.

**Comment by ezyang** [comment]
@robieta noticed that `is_complex` should get fastpathed in this way too.

**Comment by ezyang** [comment]
I think that `resize_` can also be fast-pathed (similar to @ngimel's PR to how `empty` got unified)

**Comment by dzhulgakov** [comment]
We've also discussed fast-passing `size` and `contiguous` to bypass dispatch overhead given how common they are. cc @hlu1

**Comment by ezyang** [comment]
With #49092 there is now an idiom you can do to do this:

1. Define your function with `manual_cpp_binding: True`
2. Write a manual binding with the desired fast path
3. Profit (bindings for functions are still generated)

**Comment by peterbell10** [comment]
@ezyang does that mean all that's left here is to create native operators for the manual functions listed in  the issue? If so, I could give it a go.

**Comment by ezyang** [comment]
Yes. You also need to delete the manually added code. When I devirtualized numel in https://github.com/pytorch/pytorch/pull/27294 I added a bunch of stuff, at the very least this has to be deleted, and possibly more.

@ailzhang did you already start doing this? You mentioned you would take a look.
